File: 29Items-Projects/7Algorithms/3-Dynamic-Programming/java/src/main/java/com/example/dp/ClimbingStairs.java
package com.example.dp;

import java.util.HashMap;
import java.util.Map;

public final class ClimbingStairs {
    private ClimbingStairs() {
    }

    public static Result memoized(int steps) {
        return new Result("Climbing Stairs", "memoization", solve(steps, new HashMap<>()), Map.of());
    }

    private static int solve(int steps, Map<Integer, Integer> memo) {
        if (steps <= 1) {
            return 1;
        }
        if (memo.containsKey(steps)) {
            return memo.get(steps);
        }
        int value = solve(steps - 1, memo) + solve(steps - 2, memo);
        memo.put(steps, value);
        return value;
    }

    public static Result tabulated(int steps) {
        int[] dp = new int[steps + 1];
        dp[0] = 1;
        if (steps >= 1) {
            dp[1] = 1;
        }
        for (int i = 2; i <= steps; i++) {
            dp[i] = dp[i - 1] + dp[i - 2];
        }
        return new Result("Climbing Stairs", "tabulation", dp[steps], Map.of("steps", steps));
    }

    public static Result spaceOptimized(int steps) {
        int oneBack = 1;
        int twoBack = 1;
        for (int i = 2; i <= steps; i++) {
            int current = oneBack + twoBack;
            twoBack = oneBack;
            oneBack = current;
        }
        return new Result("Climbing Stairs", "space optimized", oneBack, Map.of());
    }
}

File: 29Items-Projects/7Algorithms/3-Dynamic-Programming/python/dp_algorithms/climbing_stairs.py
from __future__ import annotations

from functools import lru_cache

from dp_algorithms.common import DPResult, require_non_negative


def memoized(steps: int) -> DPResult:
    require_non_negative(steps, "steps")

    @lru_cache(maxsize=None)
    def solve(n: int) -> int:
        return 1 if n <= 1 else solve(n - 1) + solve(n - 2)

    return DPResult("Climbing Stairs", "memoization", solve(steps))


def tabulated(steps: int) -> DPResult:
    require_non_negative(steps, "steps")
    dp = [0] * (steps + 1)
    dp[0] = 1
    if steps >= 1:
        dp[1] = 1
    for i in range(2, steps + 1):
        dp[i] = dp[i - 1] + dp[i - 2]
    return DPResult("Climbing Stairs", "tabulation", dp[steps], {"ways_by_step": dp})


def space_optimized(steps: int) -> DPResult:
    require_non_negative(steps, "steps")
    one_back, two_back = 1, 1
    for _ in range(2, steps + 1):
        one_back, two_back = one_back + two_back, one_back
    return DPResult("Climbing Stairs", "space optimized", one_back)
